### PR TITLE
build(cmake): error build on warning

### PR DESCRIPTION
**Pull request by ReenigneArcher**
## Description

This PR will error cmake on warning, causing CI to fail.

Todo:
- [x] Fix existing warnings
- [x] Add CMake option to disable -Werror

## Type of Change
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Dependency update (updates to dependencies)
- [ ] Documentation update (changes to documentation)
- [ ] Repository update (changes to repository files, e.g. `.github/...`)

## Checklist
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added or updated the in code docstring/documentation-blocks for new or existing methods/components

## Branch Updates
LizardByte requires that branches be up-to-date before merging. This means that after any PR is merged, this branch
must be updated before it can be merged. You must also
[Allow edits from maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [ ] I want maintainers to keep my branch updated

### CODE REVIEW

**Comment by ReenigneArcher** [comment]
MacOS warnings were previously fixed.

Linux builds are in pretty good shape... only failing on Debian Bookworm and flatpak.

```txt
In file included from /usr/include/c++/12.2.0/string:53,
                 from /usr/include/c++/12.2.0/stdexcept:39,
                 from /usr/include/c++/12.2.0/system_error:41,
                 from /usr/include/c++/12.2.0/bits/fs_fwd.h:35,
                 from /usr/include/c++/12.2.0/filesystem:44,
                 from /run/build/sunshine/src/config.cpp:6:
In member function ‘std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::pointer std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::_M_data() const [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>]’,
    inlined from ‘bool std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::_M_is_local() const [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>]’ at /usr/include/c++/12.2.0/bits/basic_string.h:274:23,
    inlined from ‘void std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::_M_dispose() [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>]’ at /usr/include/c++/12.2.0/bits/basic_string.h:285:18,
    inlined from ‘std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::~basic_string() [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>]’ at /usr/include/c++/12.2.0/bits/basic_string.h:795:19,
    inlined from ‘void __static_initialization_and_destruction_0(int, int)’ at /run/build/sunshine/src/config.cpp:455:3:
/usr/include/c++/12.2.0/bits/basic_string.h:234:28: error: ‘config::sunshine.config::sunshine_t::cmd.config::sunshine_t::cmd_t::name.std::__cxx11::basic_string<char>::_M_dataplus.std::__cxx11::basic_string<char>::_Alloc_hider::_M_p’ may be used uninitialized [-Werror=maybe-uninitialized]
  234 |       { return _M_dataplus._M_p; }
      |                            ^~~~
/run/build/sunshine/src/config.cpp: In function ‘void __static_initialization_and_destruction_0(int, int)’:
/run/build/sunshine/src/config.cpp:442:14: note: ‘config::sunshine’ declared here
  442 |   sunshine_t sunshine {
      |              ^~~~~~~~
In member function ‘void std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::_M_dispose() [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>]’,
    inlined from ‘std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::~basic_string() [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>]’ at /usr/include/c++/12.2.0/bits/basic_string.h:795:19,
    inlined from ‘void __static_initialization_and_destruction_0(int, int)’ at /run/build/sunshine/src/config.cpp:455:3:
/usr/include/c++/12.2.0/bits/basic_string.h:286:21: error: ‘config::sunshine.config::sunshine_t::cmd.config::sunshine_t::cmd_t::name.std::__cxx11::basic_string<char>::<anonymous>.std::__cxx11::basic_string<char>::<unnamed union>::_M_allocated_capacity’ may be used uninitialized [-Werror=maybe-uninitialized]
  286 |           _M_destroy(_M_allocated_capacity);
      |           ~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~
/run/build/sunshine/src/config.cpp: In function ‘void __static_initialization_and_destruction_0(int, int)’:
/run/build/sunshine/src/config.cpp:442:14: note: ‘config::sunshine’ declared here
  442 |   sunshine_t sunshine {
      |              ^~~~~~~~
```

Windows has some issues with vigem (and others), that I don't know how to fix.

```txt
D:\a\Sunshine\Sunshine\src\platform\windows\input.cpp: In member function 'int platf::vigem_t::alloc_gamepad_internal(const platf::gamepad_id_t&, platf::feedback_queue_t&, VIGEM_TARGET_TYPE)':
D:\a\Sunshine\Sunshine\src\platform\windows\input.cpp:277:56: error: 'VIGEM_ERROR vigem_target_ds4_register_notification(PVIGEM_CLIENT, PVIGEM_TARGET, PFN_VIGEM_DS4_NOTIFICATION, LPVOID)' is deprecated [-Werror=deprecated-declarations]
  277 |         status = vigem_target_ds4_register_notification(client.get(), gamepad.gp.get(), ds4_notify, this);
      |                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from D:\a\Sunshine\Sunshine\src\platform\windows\input.cpp:11:
D:/a/Sunshine/Sunshine/third-party/ViGEmClient/include/ViGEm/Client.h:360:48: note: declared here
  360 |         VIGEM_API VIGEM_DEPRECATED VIGEM_ERROR vigem_target_ds4_register_notification(
      |                                                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
D:\a\Sunshine\Sunshine\src\platform\windows\input.cpp: In function 'void platf::gamepad_touch(input_t&, const gamepad_touch_t&)':
D:\a\Sunshine\Sunshine\src\platform\windows\input.cpp:1601:7: error: 'pointerIndex' may be used uninitialized [-Werror=maybe-uninitialized]
 1601 |       if (pointerIndex == 0) {
      |       ^~
D:\a\Sunshine\Sunshine\src\platform\windows\input.cpp:1524:13: note: 'pointerIndex' was declared here
 1524 |     uint8_t pointerIndex;
      |             ^~~~~~~~~~~~
In file included from D:/a/_temp/msys64/mingw64/include/comdef.h:19,
                 from D:/a/Sunshine/Sunshine/src/nvenc/nvenc_d3d11.h:4,
                 from D:\a\Sunshine\Sunshine\src\platform\windows\display_vram.cpp:20:
D:/a/_temp/msys64/mingw64/include/comutil.h: In member function 'int _bstr_t::_Compare(const _bstr_t&) const':
D:/a/_temp/msys64/mingw64/include/comutil.h:272: note: '-Wmisleading-indentation' is disabled from this point onwards, since column-tracking was disabled due to the size of the code/headers
  272 |   if(m_Data==str.m_Data) return 0;
      | 
D:/a/_temp/msys64/mingw64/include/comutil.h:272: note: adding '-flarge-source-files' will allow for more column-tracking support, at the expense of compilation time and memory
cc1plus.exe: all warnings being treated as errors
mingw32-make[2]: *** [CMakeFiles\sunshine.dir\build.make:471: CMakeFiles/sunshine.dir/src/platform/windows/input.cpp.obj] Error 1
mingw32-make[2]: *** Waiting for unfinished jobs....
In file included from D:/a/_temp/msys64/mingw64/include/boost/container_hash/hash.hpp:33,
                 from D:/a/_temp/msys64/mingw64/include/boost/functional/hash.hpp:6,
                 from D:/a/_temp/msys64/mingw64/include/boost/thread/detail/thread.hpp:41,
                 from D:/a/_temp/msys64/mingw64/include/boost/thread/thread_only.hpp:22,
                 from D:/a/_temp/msys64/mingw64/include/boost/thread/thread.hpp:12,
                 from D:/a/_temp/msys64/mingw64/include/boost/log/sinks/async_frontend.hpp:37,
                 from D:/a/_temp/msys64/mingw64/include/boost/log/sinks.hpp:25,
                 from D:/a/Sunshine/Sunshine/src/logging.h:11,
                 from D:/a/Sunshine/Sunshine/src/platform/common.h:13,
                 from D:\a\Sunshine\Sunshine\src\platform\windows\display.h:14,
                 from D:\a\Sunshine\Sunshine\src\platform\windows\display_base.cpp:15:
D:/a/_temp/msys64/mingw64/include/c++/13.2.0/complex: In function 'std::basic_istream<_CharT, _Traits>& std::operator>>(basic_istream<_CharT, _Traits>&, complex<_Tp>&)':
D:/a/_temp/msys64/mingw64/include/c++/13.2.0/complex:550: note: '-Wmisleading-indentation' is disabled from this point onwards, since column-tracking was disabled due to the size of the code/headers
  550 |       if (__fail)
      | 
D:/a/_temp/msys64/mingw64/include/c++/13.2.0/complex:550: note: adding '-flarge-source-files' will allow for more column-tracking support, at the expense of compilation time and memory
```

**Comment by ReenigneArcher** [comment]
Windows issues:

- vigem depreciated

```txt
D:\a\Sunshine\Sunshine\src\platform\windows\input.cpp: In member function 'int platf::vigem_t::alloc_gamepad_internal(const platf::gamepad_id_t&, platf::feedback_queue_t&, VIGEM_TARGET_TYPE)':
D:\a\Sunshine\Sunshine\src\platform\windows\input.cpp:277:56: error: 'VIGEM_ERROR vigem_target_ds4_register_notification(PVIGEM_CLIENT, PVIGEM_TARGET, PFN_VIGEM_DS4_NOTIFICATION, LPVOID)' is deprecated [-Werror=deprecated-declarations]
  277 |         status = vigem_target_ds4_register_notification(client.get(), gamepad.gp.get(), ds4_notify, this);
      |                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from D:\a\Sunshine\Sunshine\src\platform\windows\input.cpp:11:
D:/a/Sunshine/Sunshine/third-party/ViGEmClient/include/ViGEm/Client.h:360:48: note: declared here
  360 |         VIGEM_API VIGEM_DEPRECATED VIGEM_ERROR vigem_target_ds4_register_notification(
      |                                                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

- system headers

```txt
In file included from D:/a/_temp/msys64/mingw64/include/comdef.h:19,
                 from D:/a/Sunshine/Sunshine/src/nvenc/nvenc_d3d11.h:4,
                 from D:\a\Sunshine\Sunshine\src\platform\windows\display_vram.cpp:20:
D:/a/_temp/msys64/mingw64/include/comutil.h: In member function 'int _bstr_t::_Compare(const _bstr_t&) const':
D:/a/_temp/msys64/mingw64/include/comutil.h:272: note: '-Wmisleading-indentation' is disabled from this point onwards, since column-tracking was disabled due to the size of the code/headers
  272 |   if(m_Data==str.m_Data) return 0;
      | 
D:/a/_temp/msys64/mingw64/include/comutil.h:272: note: adding '-flarge-source-files' will allow for more column-tracking support, at the expense of compilation time and memory
In file included from D:/a/_temp/msys64/mingw64/include/boost/container_hash/hash.hpp:33,
                 from D:/a/_temp/msys64/mingw64/include/boost/functional/hash.hpp:6,
                 from D:/a/_temp/msys64/mingw64/include/boost/thread/detail/thread.hpp:41,
                 from D:/a/_temp/msys64/mingw64/include/boost/thread/thread_only.hpp:22,
                 from D:/a/_temp/msys64/mingw64/include/boost/thread/thread.hpp:12,
                 from D:/a/_temp/msys64/mingw64/include/boost/log/sinks/async_frontend.hpp:37,
                 from D:/a/_temp/msys64/mingw64/include/boost/log/sinks.hpp:25,
                 from D:/a/Sunshine/Sunshine/src/logging.h:11,
                 from D:/a/Sunshine/Sunshine/src/platform/common.h:13,
                 from D:\a\Sunshine\Sunshine\src\platform\windows\display.h:14,
                 from D:\a\Sunshine\Sunshine\src\platform\windows\display_base.cpp:15:
D:/a/_temp/msys64/mingw64/include/c++/13.2.0/complex: In function 'std::basic_istream<_CharT, _Traits>& std::operator>>(basic_istream<_CharT, _Traits>&, complex<_Tp>&)':
D:/a/_temp/msys64/mingw64/include/c++/13.2.0/complex:550: note: '-Wmisleading-indentation' is disabled from this point onwards, since column-tracking was disabled due to the size of the code/headers
  550 |       if (__fail)
      | 
D:/a/_temp/msys64/mingw64/include/c++/13.2.0/complex:550: note: adding '-flarge-source-files' will allow for more column-tracking support, at the expense of compilation time and memory
```

I tried this to resolve the system header warnings, no luck.

```cmake
# ensure mingw headers are added as system headers
# Find the gcc executable
find_program(GCC_EXECUTABLE gcc)

# If found, derive the include directory
if(GCC_EXECUTABLE)
    # Assume the include directory is two levels up from the gcc binary and then in the include folder
    get_filename_component(GCC_DIR "${GCC_EXECUTABLE}" DIRECTORY)
    get_filename_component(GCC_DIR "${GCC_DIR}" DIRECTORY)
    set(GCC_INCLUDE_DIR "${GCC_DIR}/include")
    message(STATUS "GCC include directory: ${GCC_INCLUDE_DIR}")

    # Add gcc include directory as a system include directory
    include_directories(BEFORE SYSTEM "${GCC_INCLUDE_DIR}")
else()
    message(FATAL_ERROR "GCC not found")
endif()
```

File: cmake/compile_definitions/windows.cmake [hunk 1]
```diff
@@ -6,6 +6,9 @@ enable_language(RC)
 set(CMAKE_RC_COMPILER windres)
 set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -static")
 
+# gcc complains about misleading indentation in some mingw includes
+list(APPEND SUNSHINE_COMPILE_OPTIONS -Wno-misleading-indentation)
```
Comment (ReenigneArcher) [inline cmake/compile_definitions/windows.cmake:10]: Any better ways to solve the mingw warnings?

File: cmake/compile_definitions/windows.cmake [hunk 2]
```diff
@@ -28,8 +31,14 @@ file(GLOB NVPREFS_FILES CONFIGURE_DEPENDS
 include_directories(SYSTEM "${CMAKE_SOURCE_DIR}/third-party/ViGEmClient/include")
 set_source_files_properties("${CMAKE_SOURCE_DIR}/third-party/ViGEmClient/src/ViGEmClient.cpp"
         PROPERTIES COMPILE_DEFINITIONS "UNICODE=1;ERROR_INVALID_DEVICE_OBJECT_PARAMETER=650")
+set(VIGEM_COMPILE_FLAGS "")
+string(APPEND VIGEM_COMPILE_FLAGS "-Wno-unknown-pragmas ")
+string(APPEND VIGEM_COMPILE_FLAGS "-Wno-misleading-indentation ")
+string(APPEND VIGEM_COMPILE_FLAGS "-Wno-class-memaccess ")
+string(APPEND VIGEM_COMPILE_FLAGS "-Wno-unused-function ")
+string(APPEND VIGEM_COMPILE_FLAGS "-Wno-unused-variable ")
```
Comment (ReenigneArcher) [inline cmake/compile_definitions/windows.cmake:39]: Separated to reduce line length.

File: cmake/prep/options.cmake [hunk 3]
```diff
@@ -1,3 +1,5 @@
+option(BUILD_WERROR "Enable -Werror flag." ON)
```
Comment (ReenigneArcher) [inline cmake/prep/options.cmake:1]: On by default, but others can disable it if they choose.

File: cmake/compile_definitions/common.cmake [hunk 4]
```diff
@@ -3,7 +3,25 @@
...
+    # GCC 12 and higher will complain about maybe-uninitialized
+    if(CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 12)
+        list(APPEND SUNSHINE_COMPILE_OPTIONS -Wno-maybe-uninitialized)
+    endif()
+elseif(CMAKE_CXX_COMPILER_ID STREQUAL "Clang")
+    # Clang specific compile options
+
+    # Clang doesn't actually complain about this this, so disabling for now
+    # list(APPEND SUNSHINE_COMPILE_OPTIONS -Wno-uninitialized)
+endif()
```
Comment (ReenigneArcher) [inline cmake/compile_definitions/common.cmake:21]: Nothing enabled currently, but keeping the else block as a placeholder for future Clang specific items.

**Comment by ReenigneArcher** [comment]
See #2168

File: cmake/prep/options.cmake (hunk 3 above)
Comment (cgutman) [reply cmake/prep/options.cmake:1]: Personally I would default to off and just force it on for our CI. The risk is basically that somebody packages Sunshine for their distro, then the package no longer builds warning-free after the distro bumps their GCC version some time later.

As upstream, we can monitor for new warnings in new compiler versions and fix or suppress them, but I don't think we want to put that burden on downstream packagers.

Comment (ReenigneArcher) [reply cmake/prep/options.cmake]: I updated it. Only code-ql won't have this option enabled, but I don't see that as an issue.

**Pull request merged**